Allow build to explicitly set the media stack

tuncer opened the pull request:
<!-- Please describe your changes on the following line: -->
_[This is based on @asajeffrey's work in #23423, and is a rebase and continuation of that branch]_

This allows `./mach build` to override the media stack, which is currently hard-wired based on the target. To skip gstreamer, run `./mach build -d --media-stack=dummy`.

---
<!-- Thank you for contributing to Servo! Please replace each `[ ]` by `[X]` when the step is complete, and replace `___` with appropriate data: -->
- [x] `./mach build -d` does not report any errors
- [x] `./mach test-tidy` does not report any errors
- [x] These changes fix #23050 (GitHub issue number if applicable)
<!-- Either: -->
- [x] These changes do not require tests because it's build infra

<!-- Also, please make sure that "Allow edits from maintainers" checkbox is checked, so that we can help you if you get stuck somewhere along the way.-->

<!-- Pull requests that do not address these steps are welcome, but they will require additional verification as part of the review process. -->

Commit:
Allow mach build to explicitly set the media stack

You can select the media backend by building as follows:

$ ./mach build -d --media-stack=dummy
or
$ ./mach build -d --media-stack=gstreamer

Commit:
Pass feature list to build_env()

highfive commented:
Heads up! This PR modifies the following files:
 * @wafflespeanut: python/servo/command_base.py, python/servo/build_commands.py

tuncer commented:
I ran into this

    error: some components unavailable for download for channel
        nightly-2019-10-26: 'lldb-preview', 'miri', 'rls'

when trying `./mach build` locally.

tuncer commented:
Please note that I'm not a Python programmer, so it's possible the Python diff is wrong/bad.

jdm commented:
Review ping @asajeffrey.

asajeffrey commented:
This looks good to me! @bors-servo r+

bors-servo commented:
:pushpin: Commit ede3d69 has been approved by `asajeffrey`

<!-- @bors-servo r=asajeffrey ede3d69ab0e55af4a7da1150dcf7dbc4e9bd4034 -->

tuncer commented:
@jdm, @asajeffrey do you think this is sufficient to resolve #23050?

tuncer commented:
@asajeffrey, what about #23114?

asajeffrey commented:
Yes to #23050, but not #23114.

bors-servo commented:
:hourglass: Testing commit ede3d69ab0e55af4a7da1150dcf7dbc4e9bd4034 with merge 93e737d573d35db335ca9bddeef2013ca1f20e92...

bors-servo commented:
:broken_heart: Test failed - [linux-rel-css](http://build.servo.org/builders/linux-rel-css/builds/14124)

jdm commented:
```
Error running mach:

    ['test-wpt', '--release', '--processes', '24', '--total-chunks', '2', '--this-chunk', '2', '--log-raw', 'test-wpt.log', '--log-errorsummary', 'wpt-errorsummary.log', '--always-succeed']

The error occurred in code that was called by the mach command. This is either
a bug in the called code itself or in the way that mach is calling it.

You should consider filing a bug for this issue.

If filing a bug, please include the full output of mach, including this error
message.

The details of the failure are as follows:

TypeError: argument of type 'NoneType' is not iterable

  File "/home/servo/buildbot/slave/linux-rel-css/build/python/servo/testing_commands.py", line 395, in test_wpt
    ret = self.run_test_list_or_dispatch(kwargs["test_list"], "wpt", self._test_wpt, **kwargs)
  File "/home/servo/buildbot/slave/linux-rel-css/build/python/servo/testing_commands.py", line 426, in run_test_list_or_dispatch
    return correct_function(**kwargs)
  File "/home/servo/buildbot/slave/linux-rel-css/build/python/servo/testing_commands.py", line 417, in _test_wpt
    self.set_run_env(android)
  File "/home/servo/buildbot/slave/linux-rel-css/build/python/servo/command_base.py", line 596, in set_run_env
    if not android and self.needs_gstreamer_env(None, os.environ):
  File "/home/servo/buildbot/slave/linux-rel-css/build/python/servo/command_base.py", line 568, in needs_gstreamer_env
    if "media-dummy" in features:
```

Commit:
Fix needs_gstreamer_env() signature

Use iterable default value for `features`.

tuncer commented:
I'm not a Python programmer, but I think changing `None` to `[]` should make arg's default value iterable. Are these extra tests that weren't executed in the initial CI run, which was successful?

jdm commented:
If you're referring to the Community-TC check, yes, our main CI exercises many more codepaths than that one.

tuncer commented:
@jdm, ping?

jdm commented:
Redirecting ping to @asajeffrey :)

tuncer commented:
Thanks for the redirect. Am I also allowed to talk to bors-servo, or is it just for repo members?

jdm commented:
@tuncer Users who are not in the list at https://github.com/servo/saltfs/blob/4b90119f882a8024404f654dfd487429212a176e/homu/files/cfg.toml#L132-L215 will not be able to able to control @bors-servo. If you'd like us to run some commands for you we're happy to, or if you believe that being to run your own testing commands would allow you to make progress quicker then we can discuss adding you to the list of try users.

bors-servo commented:
:hourglass: Trying commit 4e33758062a86fde534b0bfbcfc22447af1e7ae9 with merge f52796ab490af2971c6da72404407a0e17e8aff7...

jdm commented:
Wow, that was... unintentional.

bors-servo commented:
:broken_heart: Test failed - [status-taskcluster](https://community-tc.services.mozilla.com/tasks/groups/L0kdxIqvSaGACeUrLh-9jg)

tuncer commented:
@jdm thanks for the explanation. I'm just curious how it works, no immediate need for bors-servo.

Do you know what the `homu-tmp` tag is? I'm asking because it's unstable/rewritten and therefore requires caution/extra-steps when running `git fetch --all` in servo.git. I can think of use cases for such a tag, so I'm, again :), just curious how servo developers handle `git fetch`.

tuncer commented:
```
error: Package `simpleservo_capi v0.0.1 (C:\Users\task_1573824755\repo\ports\libsimpleservo\capi)`
does not have these features: `media-dummy`
```
```
error: Package `simpleservo_jniapi v0.0.1 (/repo/ports/libsimpleservo/jniapi)`
does not have these features: `media-gstreamer`
```

jdm commented:
`homu-tmp` is a tag used internally by https://github.com/servo/homu/blob/4ef82e7940b3ab57ffb2bcf5bcac1dbb89212ad1/homu/main.py#L1135; I'm not sure what for. I think Servo developers tend to fetch particular branches like master rather than --all.

tuncer commented:
Maybe `media-dummy` and `media-gstreamer` features are missing in `libsimpleservo/Cargo.toml`?

jdm commented:
Yes, that would make sense.

Commit:
simpleservo: add media-{dummy,gstreamer} features

tuncer commented:
In capi and jniapi, should it be `media-foo = ["simpleservo/servo-media-foo"]` or `media-foo = ["libservo/servo-media-foo"]`?

tuncer commented:
Looking at the Cargo error, I'm not sure that's what was missing.

jdm commented on ports/libsimpleservo/capi/Cargo.toml:
```diff
@@ -31,6 +31,8 @@ canvas2d-azure = ["simpleservo/canvas2d-azure"]
 debugmozjs = ["simpleservo/debugmozjs"]
 default = ["webdriver", "max_log_level"]
+media-dummy = ["libservo/servo-media-dummy"]
+media-gstreamer = ["libservo/servo-media-gstreamer"]
```
Both of these crates depends on simpleservo, not libservo.

tuncer replied:
It's possible I'm confused by the project structure. `components/servo/Cargo.toml` defines the two features `servo-media-dummy` and `servo-media-gstreamer`, which are selected implicitly when enabling libsimpleservo's `media-dummy` or `media-gstreamer` feature. Same for the glutin port. And depending on which feature is selected, one of the `media_platform` modules is built conditionally in libservo.

What am I missing (that would help resolve the issue)?

tuncer replied:
`capi` and `jniapi` depend on the common `libsimpleservo/api` crate, and that one has the expected `libservo` dependency. Hmm...

tuncer replied:
I'll add `media-dummy` and `media-gstreamer` features to `libsimpleservo/api`, to be used by `capi` and `jniapi`, since `api` is the common crate which depends on `libservo`.

jdm commented:
" Feature `media-dummy` requires a feature of `libservo` which is not a dependency"

tuncer commented:
> " Feature `media-dummy` requires a feature of `libservo` which is not a dependency"

We cannot depend on libservo like `components/glutin/Cargo.toml` does, can we?

tuncer commented:
All checks (Community-TC) have passed and I'm done for today

tuncer commented:
Can we queue it for full CI?

jdm commented:
@bors-servo try

bors-servo commented:
:hourglass: Trying commit 0b4e5eeb871d8cc6c17adc5859eb15a9a9768514 with merge 501073c1e0ecb737422170e7c1ce634e3f49da54...

tuncer commented:
Thanks, _Linux x64: Docs + check_ is the only failure left. Time to fix that one.

bors-servo commented:
:broken_heart: Test failed - [status-taskcluster](https://community-tc.services.mozilla.com/tasks/groups/e7EYIXEAS76qbkAYE3ygcQ)

tuncer commented:
Any idea what's missing when running `cargo doc`?

Linux x64: Docs + check
```
    Checking layout_thread_2013 v0.0.1 (/repo/components/layout_thread)
 Documenting layout_thread_2013 v0.0.1 (/repo/components/layout_thread)
error[E0433]: failed to resolve: use of undeclared type or module `media_platform`
   --> components/servo/lib.rs:332:13
    |
332 |             media_platform::init();
    |             ^^^^^^^^^^^^^^ use of undeclared type or module `media_platform`
error[E0433]: failed to resolve: use of undeclared type or module `media_platform`
   --> components/servo/lib.rs:972:5
    |
972 |     media_platform::init();
    |     ^^^^^^^^^^^^^^ use of undeclared type or module `media_platform`
warning: ignoring -C extra-filename flag due to -o flag
error: aborting due to 2 previous errors
For more information about this error, try `rustc --explain E0433`.
error: could not compile `libservo`.
warning: build failed, waiting for other jobs to finish...
error: build failed
```

tuncer commented:
Hmm, not sure why `media_platform` is undefined when running `cargo doc`

Linux x64: Docs + check
```
    Checking layout_thread_2013 v0.0.1 (/repo/components/layout_thread)
 Documenting layout_thread_2013 v0.0.1 (/repo/components/layout_thread)
error[E0433]: failed to resolve: use of undeclared type or module `media_platform`
   --> components/servo/lib.rs:332:13
    |
332 |             media_platform::init();
    |             ^^^^^^^^^^^^^^ use of undeclared type or module `media_platform`
error[E0433]: failed to resolve: use of undeclared type or module `media_platform`
   --> components/servo/lib.rs:972:5
    |
972 |     media_platform::init();
    |     ^^^^^^^^^^^^^^ use of undeclared type or module `media_platform`
warning: ignoring -C extra-filename flag due to -o flag
error: aborting due to 2 previous errors
For more information about this error, try `rustc --explain E0433`.
error: could not compile `libservo`.
warning: build failed, waiting for other jobs to finish...
error: build failed
```

tuncer commented:
We also need to teach `components/servo/Cargo.toml` to include `gstreamer` only if `media-stack==gstreamer`.

What we have right now is

```
[target.'cfg(any(target_os = "android", target_arch = "x86_64",
target_os = "windows"))'.dependencies]
gstreamer = "0.14.5"
```

and we cannot do
```
[target.'cfg(feature = "servo-media-gstreamer")'.dependencies]
gstreamer = "0.14.5"
```

Maybe we can do this in `[features]` and remove the (currently) mandatory `gstreamer` dependency.

Commit:
Make gstreamer(-sys) optional

tuncer commented:
Fixed known issues, including accidentally still mandatory gstreamer (crate) dependency. I have also (locally) built a `media_dummy` servo executable that starts and loads a remote website. Let's queue it for full CI.

tuncer commented:
Before we merge this, should we reconsider the flag's name, or is the already inconsistent naming (see `./mach build --help`) reason to keep `--media-stack=` instead of, say, renaming to `--without-gstreamer` or `--with-media-stack=`?

tuncer commented:
Still waiting for `./mach doc` to finish, which, for some reason, is checking+documenting all crates, not just the immediate/1st-level crates. Is this normal? Put differently, should `./mach doc` really process `serde_urlencoded` and __everything__ else servo depends on?

Commit:
Implement `./mach doc --media-stack={dummy|gstreamer}`

tuncer commented:
Fixed all known issues and successfully loaded a remote website with a locally built media-dummy servo executable.

tuncer commented:
Can we queue it for full CI?

CYBAI commented:
@bors-servo try

bors-servo commented:
:hourglass: Trying commit b4b045c3c03c78a47314280457e58c07fa324c51 with merge 25a4b20300f006afc9ac60641472e15149ac2814...

bors-servo commented:
:broken_heart: Test failed - [status-taskcluster](https://community-tc.services.mozilla.com/tasks/groups/f5LcFN0YSxWZC2CdHtcq8Q)

tuncer commented:
I don't know what's going on with _[macOS x64: WPT chunk 4 / 6](https://community-tc.services.mozilla.com/tasks/TBee4gx4QDOY7TKThgd-aw/runs/0)_.

tuncer commented:
I successfully ran `./mach doc --media-stack=dummy` yesterday and the selection code is the same as in `./mach build`. Thus, I don't know why _[Linux x64: Docs + check](https://community-tc.services.mozilla.com/tasks/KYXGgDBcQD-8LhBlmBRfPw/runs/0)_ fails to conditionally build the `media_platform` mod in `components/servo/lib.rs`.

tuncer commented:
Thanks, @CYBAI.

tuncer commented:
Rebased, since 591d7ad6c18875ed2f5cc3dc4f54e0b2b53c50fc touches `build_commands.py`.

jdm commented:
It's `./mach check` that is failing to build according to the log.

Commit:
Implement `./mach check --media-stack`

While at it, extract --media-stack flag into command_base.py.

tuncer commented:
Extracted `--media-stack` handling into `command_base.py` and implemented media stack selection in `./mach check`, too. Let's try bors again, though we haven't identified the macOS test failure yet, have we?

tuncer commented:
By the way, how do you invoke `./mach doc`? Do you pass `--no-deps`, or how else do you prevent documenting all dependencies?

jdm commented:
The macos test failure is #23290.

jdm commented:
https://github.com/servo/servo/blob/5c92fd84cab4d4e214a8e16042812a38480817ba/etc/taskcluster/decision_task.py#L220 is the command to invoke mach doc on CI.

Commit:
Allow media_stack in run_cargo_build_like_command

tuncer commented:
bors-servo try :question:

tuncer commented:
@jdm we now support `--media-stack` in `./mach` `build`, `check`, `doc`, and `test-unit`. Anything else before we retry a full CI?

tuncer commented:
I'll check if we can extract more of the repeated code into a shared function. I'm not a Python programmer, so no promises...

Commit:
Extract media_stack pick into CommandBase

tuncer commented:
@jdm since 8512212, we support `--media-stack` in `./mach` `build`, `check`, and `doc`. Plus, I've extracted more code into `CommandBase`.

Anything else before a full CI? Do you know of other commands that have to pick and submit the media stack?

tuncer commented:
@CYBAI, can we trigger full CI?

jdm commented:
@bors-servo try

bors-servo commented:
:hourglass: Trying commit 81c439a8990caf498bd3318cbe5f51c33830f7e3 with merge 46ab840a00e9e799759f702d001e9a524062c16c...

tuncer commented:
Thanks, @jdm!

tuncer commented:
Once this is merged, I'd like to add the dummy variant/artifact to https://github.com/servo/download.servo.org. Do you know where the build script is that uploads nightly artifacts?

jdm commented:
All of the tasks in https://github.com/servo/servo/blob/master/etc/taskcluster/decision_task.py that include `mach upload-nightly`, and that command's implementation lives in https://github.com/servo/servo/blob/master/python/servo/package_commands.py#L579.

tuncer commented:
Thanks, I'll check it out. Should I extend it in this pull request or open a separate ticket, later?

tuncer commented:
Draft patch: https://github.com/servo/download.servo.org/pull/54

jdm commented:
A separate ticket, please. I'm not yet convinced that we want to go to the effort of building and distributing non-gstreamer nightly builds.

tuncer commented:
I agree, let's not delay this diff unnecessarily. BTW, what's the status of https://github.com/servo/download.servo.org/pull/53?

jdm commented:
The status was forgotten, and the new status is merged.

tuncer commented:
Here's a first draft: #24829

bors-servo commented:
:sunny: Test successful - [linux-rel-css](http://build.servo.org/builders/linux-rel-css/builds/14297), [linux-rel-wpt](http://build.servo.org/builders/linux-rel-wpt/builds/14363), [status-taskcluster](https://community-tc.services.mozilla.com/tasks/groups/PU5YSeNQQwyMOkxkirxkJg)
State: approved= try=True

tuncer commented:
@jdm @asajeffrey all green. Anything else left to do before the final review/approval?

asajeffrey reviewed (approved):
This is a big improvement, thanks for taking this over!

jdm commented:
@bors-servo r=asajeffrey

bors-servo commented:
:pushpin: Commit 81c439a has been approved by `asajeffrey`

<!-- @bors-servo r=asajeffrey 81c439a8990caf498bd3318cbe5f51c33830f7e3 -->

bors-servo commented:
:hourglass: Testing commit 81c439a8990caf498bd3318cbe5f51c33830f7e3 with merge 10a63cc9d41acb535353436cf276eac5fdebb971...

bors-servo commented:
:sunny: Test successful - [linux-rel-css](http://build.servo.org/builders/linux-rel-css/builds/14316), [linux-rel-wpt](http://build.servo.org/builders/linux-rel-wpt/builds/14382), [status-taskcluster](https://community-tc.services.mozilla.com/tasks/groups/Wi3kGPmWR-O2gWyytlnMhw)
Approved by: asajeffrey
Pushing 10a63cc9d41acb535353436cf276eac5fdebb971 to master...
